File: 383-ransom-note/383-ransom-note.py
class Solution(object):
    def canConstruct(self, ransomNote, magazine):
        # Time Complexity : O(n+m), where n, m is the length of 'ransomNote' and 'magazine' strings
        # Space Complexity: O(m), where m is the 'magazine' string
        latters_counter = Counter(magazine)
        
        for latter in ransomNote:
            if latters_counter[latter] == 0 or latter not in latters_counter:
                return False
            latters_counter[latter] -= 1
        
        return True

File: 383-ransom-note/383-ransom-note.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function(ransomNote, magazine) {
    const leters_counter = {}
    for(let i=0; i<magazine.length; i++){
        let letter = magazine[i]
        if (letter in leters_counter){
            leters_counter[letter] += 1
        }else{
            leters_counter[letter] = 0
        }
    }
    
    for (let letter of ransomNote){
        if (!leters_counter[letter]){
            return false
        }
        leters_counter[letter]--
    }
    return true
};

var canConstruct = function(ransomNote, magazine) {
    const map = {};
    for(let letter of magazine) {
        if (!map[letter]) {
            map[letter] = 0;
        }
        map[letter]++;
    }
    
    for(let letter of ransomNote) {
        if(!map[letter]) {
            return false;
        } 
        map[letter]--;
    }
    return true;
};